policy/engine.go

```diff
@@ -0,0 +1,784 @@
+	}
+
+	// Only allow asterisk at the start of the domain; we don't allow them as part of a domain label or as a (sub)domain label (currently)
+	if strings.LastIndex(domain, "*") > 0 {
+		return false, nil
+	}
+
+	// Don't allow constraints with empty labels in any position
+	if strings.Contains(constraint, "..") {
+		return false, nil
+	}
+
+	domainLabels, ok := domainToReverseLabels(domain)
+	if !ok {
+		return false, fmt.Errorf("cannot parse domain %q", domain)
+	}
+
+	// RFC 5280 says that a leading period in a domain name means that at
+	// least one label must be prepended, but only for URI and email
+	// constraints, not DNS constraints. The code also supports that
+	// behavior for DNS constraints. In our adaptation of the original
+	// Go stdlib x509 Name Constraint implementation we look for exactly
+	// one subdomain, currently.
+
+	mustHaveSubdomains := false
+	if constraint[0] == '.' {
+		mustHaveSubdomains = true
+		constraint = constraint[1:]
+	}
+
+	constraintLabels, ok := domainToReverseLabels(constraint)
+	if !ok {
+		return false, fmt.Errorf("cannot parse domain constraint %q", constraint)
+	}
+
+	// fmt.Println(mustHaveSubdomains)
+	// fmt.Println(constraintLabels)
+	// fmt.Println(domainLabels)
+
+	expectedNumberOfLabels := len(constraintLabels)
+	if mustHaveSubdomains {
+		// we expect exactly one more label if it starts with the "canonical" x509 "wildcard": "."
+		// in the future we could extend this to support multiple additional labels and/or more
+		// complex matching.
+		expectedNumberOfLabels++
+	}
+
+	if len(domainLabels) != expectedNumberOfLabels {
+		return false, nil
+	}
+
+	for i, constraintLabel := range constraintLabels {
+		if !strings.EqualFold(constraintLabel, domainLabels[i]) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// SOURCE: https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/crypto/x509/verify.go
+func matchIPConstraint(ip net.IP, constraint *net.IPNet) (bool, error) {
+
+	// TODO(hs): this is code from Go library, but I got some unexpected result:
+	// with permitted net 127.0.0.0/24, 127.0.0.1 is NOT allowed. When parsing 127.0.0.1 as net.IP
+	// which is in the IPAddresses slice, the underlying length is 16. The contraint.IP has a length
+	// of 4 instead. I currently don't believe that this is a bug in Go now, but why is it like that?
+	// Is there a difference because we're not operating on a sans []string slice? Or is the Go
+	// implementation stricter regarding IPv4 vs. IPv6? I've been bitten by some unfortunate differences
+	// between the two before (i.e. IPv4 in IPv6; IP SANS in ACME)
+	// if len(ip) != len(constraint.IP) {
+	// 	return false, nil
+	// }
+
+	// for i := range ip {
+	// 	if mask := constraint.Mask[i]; ip[i]&mask != constraint.IP[i]&mask {
+	// 		return false, nil
+	// 	}
+	// }
+
+	contained := constraint.Contains(ip) // TODO(hs): validate that this is the correct behavior; also check IPv4-in-IPv6 (again)
+
+	return contained, nil
+}
```

Comment:
Not sure if relevant but there's also IP.To4 and IP.To16, https://pkg.go.dev/net@go1.17.6#IP.To4